numerical_methods/fast_fourier_transform.cpp

```diff
@@ -0,0 +1,163 @@
+        cos(2 * pi / n), sin(2 * pi / n));  /// Calculating value of omega
+
+    auto *pe = new std::complex<double>[n / 2];  /// Coefficients of even power
+
+    auto *po = new std::complex<double>[n / 2];  /// Coefficients of odd power
+
+    int k1 = 0, k2 = 0;
+    for (int j = 0; j < n; j++) {
+        if (j % 2 == 0) {
+            pe[k1++] = p[j];  /// Assigning values of even Coefficients
+
+        } else
+            po[k2++] = p[j];  /// Assigning value of odd Coefficients
+    }
+
+    std::complex<double> *ye =
+        FastFourierTransform(pe, n / 2);  /// Recursive Call
+
+    std::complex<double> *yo =
+        FastFourierTransform(po, n / 2);  /// Recursive Call
+
+    auto *y = new std::complex<double>[n];  /// Final value representation list
+
+    k1 = 0, k2 = 0;
+
+    for (int i = 0; i < n / 2; i++) {
+        y[i] =
+            ye[k1] + pow(om, i) * yo[k2];  /// Updating the first n/2 elements
+        y[i + n / 2] =
+            ye[k1] - pow(om, i) * yo[k2];  /// Updating the last n/2 elements
+
+        k1++;
+        k2++;
+    }
+ 
+    if(n!=2){
+     
+        delete[] pe;
+        delete[] po;
+        
+    }
+
+    delete[] ye;  /// Deleting dynamic array ye
+    delete[] yo;  /// Deleting dynamic array yo
+    return y;
+}
+
+}  // namespace numerical_methods
+
+/**
+ * @brief Self-test implementations
+ * @details
+ * Declaring two test cases and checking for the error
+ * in predicted and true value is less than 0.000000000001.
+ * @returns void
+ */
+static void test() {
+    /* descriptions of the following test */
+
+    auto *t1 = new std::complex<double>[2];  /// Test case 1
+    auto *t2 = new std::complex<double>[4];  /// Test case 2
+
+    t1[0] = {1, 0};
+    t1[1] = {2, 0};
+    t2[0] = {1, 0};
+    t2[1] = {2, 0};
+    t2[2] = {3, 0};
+    t2[3] = {4, 0};
+
+    uint8_t n1 = 2;
+    uint8_t n2 = 4;
+    std::vector<std::complex<double>> r1 = {
+        {3, 0}, {-1, 0}};  /// True Answer for test case 1
+
+    std::vector<std::complex<double>> r2 = {
+        {10, 0}, {-2, -2}, {-2, 0}, {-2, 2}};  /// True Answer for test case 2
+
+    std::complex<double> *o1 = numerical_methods::FastFourierTransform(t1, n1);
+    std::complex<double> *o2 = numerical_methods::FastFourierTransform(t2, n2);
+
+    for (uint8_t i = 0; i < n1; i++) {
+        assert((r1[i].real() - o1->real() < 0.000000000001) &&
+               (r1[i].imag() - o1->imag() <
+                0.000000000001));  /// Comparing for both real and imaginary
+                                   /// values for test case 1
+        o1++;
+    }
+
+    for (uint8_t i = 0; i < n2; i++) {
+        assert((r2[i].real() - o2->real() < 0.000000000001) &&
+               (r2[i].imag() - o2->imag() <
+                0.000000000001));  /// Comparing for both real and imaginary
+                                   /// values for test case 2
+        o2++;
+    }
+    
+    
+    delete[] t1;
+    delete[] t2;
+

```

Comment:
```suggestion
    delete [] (o1 - n1);
    delete [] (o2 - n2);
```
This will technically fix the error.
You are actually returning  array of std::complex<double> from a function and incrementing returned pointers by `n1 and n2` therefore when you free  the memory location you have to decrement the pointer by same amount.

Reply:
Probably a better way to do the same thing would be create another pointer before incrementing o1 and o2 and free the location using those newly created pointers. Code will look more clean, readable and easy understandable.